src/shared/ThemeContext/ThemeContext.tsx

```diff
@@ -0,0 +1,47 @@
+import React, { createContext, FC, ReactNode, useEffect, useState } from 'react'
+
+export enum Theme {
+  LIGHT = 'light',
+  DARK = 'dark',
+}
+
+interface ThemeContextProps {
+  theme: Theme
+  setTheme: (theme: Theme) => void
+}
+
+const defaultThemeContext: ThemeContextProps = {
+  theme: Theme.LIGHT,
+  setTheme: () => {},
+}
+
+export const ThemeContext =
+  createContext<ThemeContextProps>(defaultThemeContext)
+
+interface ThemeContextProviderProps {
+  children: ReactNode
+}
+
+export const ThemeContextProvider: FC<ThemeContextProviderProps> = ({
+  children,
+}) => {
+  const currentTheme = (localStorage.getItem('theme') as Theme) || Theme.LIGHT
+  const [theme, setTheme] = useState<Theme>(currentTheme)
+
+  useEffect(() => {
+    document.body.classList.remove(Theme.LIGHT, Theme.DARK)
+    document.body.classList.add(theme)
+
+    localStorage.setItem('theme', theme)
+  }, [theme])
```

Comment:
Can you replace this effect with just some `if` statements and only set the values if they're different then the current values. In theory you shouldn't need an effect here iirc.

Reply:
Hm so I want this to run the first time the page is rendered, and also any subsequent time `theme` changes. Keeping it in `useEffect` will help prevent it re-running unexpectedly, so left it here this time!

Reply:
Hrrm this still smells to me a bit. Re-reading the code, you're trying to use this effect in response to an event, which is something we don't want to happen, you can read more about it [here](https://react.dev/learn/you-might-not-need-an-effect)

I would change this up slightly so it would look something along the lines of:

```tsx
const [theme, setTheme] = useState<Theme>(currentTheme)
const initialRender = useRef(true)

if (initialRender) {
  document.body.classList.remove(Theme.LIGHT, Theme.DARK)
  document.body.classList.add(theme)

  localStorage.setItem('theme', theme)
  initialRender.current = false
}

const handleTheme = useCallback(() => {
  document.body.classList.remove(Theme.LIGHT, Theme.DARK)
  document.body.classList.add(theme)
  localStorage.setItem('theme', theme)
  setTheme((oldTheme) => oldTheme === Theme.LIGHT ? Theme.DARK : Theme.LIGHT))
}, [theme])
// I think that's all that would go in the dep array, lint rule can verify
```

Reply:
Thanks for following up!
I made the suggested change. I agree this way makes it more explicit what is expected to happen when `setTheme` is called.

Will read up some more on that doc you sent over - thanks for that! My initial thought was this would fall under this use case `Use Effects only for code that should run because the component was displayed to the user` but since we're having things happen when the user selects a theme with the ThemeToggle, it does fall into the "in response to an event" camp. 
Cool discussion - thanks for your help!